eval/bundled/epm.elv.go

```diff
@@ -1,111 +1,279 @@
-        echo $pkg >> $-installed
-    }
+fn dest [pkg]{
+  put $-lib-dir/$pkg
 }
 
-fn -get-url [pkg]{
-    put https://$pkg
+fn is-installed [pkg]{
+  put ?(test -e (dest $pkg))
 }
 
-fn -install-one [pkg]{
-    dest = $-data-dir/lib/$pkg
-    if ?(test -e $dest) {
-        -error 'Package '$pkg' already exists locally.'
-        return
-    }
-    -info 'Installing '$pkg
-    mkdir -p $dest
-    git clone (-get-url $pkg) $dest
-    add-installed $pkg
+fn -package-domain [pkg]{
+  splits / $pkg | take 1
 }
 
-fn install [@pkgs]{
-    if (eq $pkgs []) {
-        -error 'Must specify at least one package.'
-        return
+fn -package-without-domain [pkg]{
+  joins / [(splits / $pkg | drop 1)]
+}
+
+# Known method handlers. Each entry is indexed by method name (the
+# value of the "method" key in the domain configs), and must contain
+# two keys: install and upgrade, each one must be a closure that
+# received two arguments: package name and the domain config entry
+-method-handler = [
+  &git= [
+    &src= [pkg dom-cfg]{
+      put $dom-cfg[protocol]"://"$pkg
     }
-    for pkg $pkgs {
-        -install-one $pkg
+
+    &install= [pkg dom-cfg]{
+      dest = (dest $pkg)
+      -info "Installing "$pkg
+      mkdir -p $dest
+      git clone ($-method-handler[git][src] $pkg $dom-cfg) $dest
     }
-}
 
-fn installed {
-    if ?(test -f $-installed) {
-        cat $-installed
+    &upgrade= [pkg dom-cfg]{
+      dest = (dest $pkg)
+      -info "Updating "$pkg
+      git -C $dest pull
+    }
+  ]
+
+  &rsync= [
+    &src= [pkg dom-cfg]{
+      put $dom-cfg[location]/(-package-without-domain $pkg)/
+    }
+
+    &install= [pkg dom-cfg]{
+      dest = (dest $pkg)
+      pkgd = (-package-without-domain $pkg)
+      -info "Installing "$pkg
+      rsync -av ($-method-handler[rsync][src] $pkg $dom-cfg) $dest
     }
-}
 
-fn -upgrade-one [pkg]{
-    dest = $-data-dir/lib/$pkg
-    if (not ?(test -d $dest)) {
-        -error 'Package '$pkg' not installed locally.'
+    &upgrade= [pkg dom-cfg]{
+      dest = (dest $pkg)
+      pkgd = (-package-without-domain $pkg)
+      if (not (is-installed $pkg)) {
+        -error "Package "$pkg" is not installed."
         return
+      }
+      -info "Updating "$pkg
+      rsync -av ($-method-handler[rsync][src] $pkg $dom-cfg) $dest
     }
-    -info 'Upgrading package '$pkg
-    git -C $dest pull
+  ]
+]
+
+# Return the filename of the domain config file for the given domain
+# (regardless of whether it exists)
+fn -domain-config-file [dom]{
+  put $-lib-dir/$dom/epm-domain.cfg
 }
 
-fn upgrade [@pkgs]{
-    if (eq $pkgs []) {
-        pkgs = [(installed)]
-        -info 'Upgrading all installed packages'
-    }
-    for pkg $pkgs {
-        -upgrade-one $pkg
+# Return the filename of the metadata file for the given package
+# (regardless of whether it exists)
+fn -package-metadata-file [pkg]{
+  put (dest $pkg)/metadata.json
+}
+
+# Read the domain config file for a given domain. If the file does not
```

Comment:
Not caching means that this function can now be rewritten as one that outputs the config instead of storing it.